README.md
# Compute Anchor

![](https://img.shields.io/travis/com/teeeemoji/compute-anchor.svg)

This is a small package, listening `scroll` or `wheel` or `touchmove` event of the given DOM element, for compute the most compatible anchor in AnchorList.
And its 

## Getting Started

### Prerequisites

This package support both pc and mobile.

### Installing

```shell
$ npm install compute-anchor
```

### Usage

```javascript
let dispose = computeAnchor({// add event listener on element
    element: document.getElementById('container'),
    anchors: ['title', 'getting start', 'installing', 'usage'],
    callback: anchor => console.log(anchor),
    offset: {
      top: 80,
      bottom: 0
    }
})

dispose()// remove event listener on element
```

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### computeAnchor

[src/compute-anchor.js:26-33](https://git@github.com/:teeeemoji/compute-anchor/blob/fa70417ed30c70c140592850b58c99fc10c2446e/src/compute-anchor.js#L26-L33 "Source code on GitHub")

This function accept an element and some other options, to start listening element's event for compute anchor

#### Parameters

-   `options`  {Object}
    -   `options.element`  {DOMNode}
           which dom element contains all anchors and its scrollable (optional, default `window`)
    -   `options.events`  {String\[]}
           which events will add on `options.elements` (optional, default `['scroll','resize']`)
    -   `options.anchors`  {String\[]}
           all anchors you care about, and it just anchor name, don't start with `#`
    -   `options.callback`  {Function}
           callback will invoke when compute handler returns a new anchor
           if compute handler returns the same result, callback will not be invoked
           `callback(anchor){}`
    -   `options.defaultAnchor`  {String}
           if can't find any fit anchor, callback will receive defaultAnchor (optional, default `''`)
    -   `options.offset`  {Number\[]}
           offset[0] is top offset, offset[1] is bottom offset
           when your input element has some fixed top or fixed bottom children, options.offset if useful (optional, default `[0,0]`)
    -   `options.wait`  {Number}
           wait is time in MS for eventListener wait time (optional, default `20`)

Returns **[Function](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function)** the function to remove all eventListener on element

## Running the tests

TODO

## Contributing

There's no other contributor

## Authors

teeeemoji <mailto:teeeemoji@163.com>

## License

This project is licensed under the MIT License - see the [LICENSE](./LICENSE) file for details 

### Task

## important

-   write test
-   publish this package

## others

-   friendly way to write demo


CHANGELOG.md
<a name="1.0.2"></a>
## [1.0.2](https://github.com/teeeemoji/compute-anchor/compare/v1.0.1...v1.0.2) (2019-07-18)


### Features

* add events in options ([82d6e65](https://github.com/teeeemoji/compute-anchor/commit/82d6e65))
* add mobile support for touch-move event ([21a6fad](https://github.com/teeeemoji/compute-anchor/commit/21a6fad))
* update demo to adapt mobile screen ([f264450](https://github.com/teeeemoji/compute-anchor/commit/f264450))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/teeeemoji/compute-anchor/compare/79a3c5b...v1.0.1) (2019-07-08)


### Bug Fixes

* update babel config ([79a3c5b](https://github.com/teeeemoji/compute-anchor/commit/79a3c5b))





src/utils.js
import compute from "./compute";
import debounce from "lodash/debounce";

//Returns true if it is a DOM element, This code is tested in FF3, IE7, Chrome 1 and Opera 9
function isElement(o) {
  return typeof HTMLElement === "object"
    ? o instanceof HTMLElement
    : o &&
        typeof o === "object" &&
        o !== null &&
        o.nodeType === 1 &&
        typeof o.nodeName === "string";
}

function getElement(ele) {
  if (!ele) {
    return window;
  }
  let element = typeof ele === "function" ? ele() : ele;
  if (isElement(element)) {
    return element;
  }
  throw new Error("[compute anchor]: input element is not valid DOM element");
}

function mobilecheck() {
  let check = false;
  (function(a) {
    if (
      /(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|mobile.+firefox|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows ce|xda|xiino/i.test(
        a
      ) ||
      /1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/i.test(
        a.substr(0, 4)
      )
    ) {
      check = true;
    }
  })(navigator.userAgent || navigator.vendor || window.opera);
  return check;
}

const isMobile = mobilecheck();

const EMPTY_OPTIONS = {
  element: null,
  handler: null,
  defaultAnchor: "",
  events: [],
  anchors: [],
  callback: anchor => {},
  offset: [0, 0], // [top, bottom]
  wait: 20 // ms
};

export function initOptions(options) {
  let eventName;
  let element = getElement(options.element);
  if (isMobile) {
    eventName = "touchmove";
  } else if (element !== window) {
    eventName = "wheel";
  } else {
    eventName = "scroll";
  }
  return {
    ...EMPTY_OPTIONS,
    ...options,
    element,
    events:
      options.events && options.events.length
        ? options.events
        : ["resize", eventName],
    offset: {
      top: (options.offset && options.offset.top) || 0,
      bottom: (options.offset && options.offset.bottom) || 0
    }
  };
}

export function makeHandler(options) {
  let lastAnchor = options.defaultAnchor;
  return function(e) {
    let anchor = compute(options);
    if (anchor !== lastAnchor) {
      lastAnchor = anchor;
      options.callback(anchor);
    }
  };
}

export function makeDebounceHandler(options, handler) {
  return debounce(handler, options.wait);
}

export function listen(options, handler) {
  console.log("listen:", options.events);
  options.events.forEach(eventName =>
    options.element.addEventListener(eventName, handler)
  );
}

export function dispose(options, handler) {
  options.events.forEach(eventName =>
    options.element.removeEventListener(eventName, handler)
  );
}
